### PR TITLE
Using Polynomials in the module holonomic and Implement conversion to Hypergeometric.

### PR DESCRIPTION
**Pull request by shubhamtibra**
For now I have changed the method `_normalize` to use `polys` thus making it more robust.
- [x] Use only Polynomials
- [x] Converting Hypergeometric Functions

### CODE REVIEW

File: sympy/holonomic/holonomic.py [hunk 1]
```diff
@@ -674,6 +707,137 @@ def __mul__(self, other):
...
+    for i in range(len(b)):
+        r2 = r2 * (x * Dx + b[i] - 1)
+    sol = r1 - r2
+
+    simp = hyperexpand(func)
+
+    if isinstance(simp, Infinity) or isinstance(simp, NegativeInfinity):
+        return HolonomicFunction(sol, x).composition(z)
+
+    def _find_conditons(simp, x, x0, order):
```
Comment (shubhamtibra) [inline sympy/holonomic/holonomic.py]: To find Initial Conditions for the Holonomic Function. Just substitutes the point x0 in the hypergeometric function. This is the case where the hypergeometric function can be computed symbolically using `hyperexpand`. I will also add for the case when the function is not know symbolically using `evalf()` .

Comment (jksuom) [inline sympy/holonomic/holonomic.py]: Sorry, I intended to question the spelling.

Comment (shubhamtibra) [inline sympy/holonomic/holonomic.py]: It's Alright. I'll rectify that in the next commit.

**Comment by shubhamtibra** [comment]
There seems to be a problem in finding Initial Conditions when the Hypergeometric Function can't be computed symbolically using `hyperexpand`.
Let's say I want to convert `f(x) = hyper([1, 1, 1], [S(3)/2], x**2)`. Then to find f'(0), I will first find `f.diff(x, 1)` which gives `f'(x) = 4*x*hyper((2, 2, 2), (5/2,), x**2)/3` now substituting `0` in this gives `0` due to the term `x` in `f'(x)`. But this is not necessarily `0` because the hypergeometric term may have a `x` in the denominator and `f'(0)` will be non zero then.

**Comment by jksuom** [comment]
The power series expansion of `f(x) = hyper([1, 1, 1], [S(3)/2], x**2)` has only even powers. Hence all odd derivatives vanish at 0.

**Comment by shubhamtibra** [comment]
You are right. Thanks @jksuom . I tried it on a different example `hyper([1], [S(3)/2], x**2).diff(x, 1).subs(x, 0).evalf()` gives `0` and `hyperexpand(hyper([1], [S(3)/2], x**2)).diff(x, 1).subs(x, 0).evalf()` gives `nan`. The answer should be same, right?

**Comment by jksuom** [comment]
I would expect so. But I forget what `hyperexpand` would do in this case. Maybe it leads to a less tractable form.

**Comment by shubhamtibra** [comment]
Yes `hyperexpand` tries to find the Hypergeometric Function. here `hyperexpand(hyper([1], [S(3)/2], x**2))` will give the function `sqrt(pi)*exp(x**2)*erf(x)/(2*x)`.

**Comment by shubhamtibra** [comment]
What should I do next ? Using only Polynomials internally can't be done right now as our linear system solver doesn't support `DMP` and `DMF` objects.

**Comment by jksuom** [comment]
It might not be too hard to work around the restrictions of the linear system solver. The actual workhorse of `gauss_jordan_solve` is `rref`. It seems that `rref` would not call `sympify` or `simplify`, so it could probably be used to solve the linear system with `DMF` objects to reduced row-echelon form. Some work would be needed to write out the final solution.

**Comment by shubhamtibra** [comment]
When I tried creating a Matrix from elements of `DMF` it throws errors. What should I do for that? I also  have tried subclassing as we discussed on gitter but that also wasn't successful.

**Comment by jksuom** [comment]
Did `rref` not work with the subclassed matrix?

**Comment by shubhamtibra** [comment]
I will be looking into it. Using `rref` directly instead of `gauss_jordan_solve`.

**Comment by shubhamtibra** [comment]
I was working on it and defined the solver to work with `DMP` and `DMF`. One problem is `diff()` is not defined on `DMF` objects. So How do I differentiate these rational functions.

**Comment by jksuom** [comment]
To differentiate rational functions you have to define a function imitating the elementary rule. The numerator and denominator are first considered separately. The derivatives of (univariate) `DMP`s are easily constructed directly without passing to expressions. (There is  a method `dup_diff` in `polys/densetools` that handles also higher derivatives..) Then the derivative of `p/q` is obtained from the standard formula `(p'q - q'p)/q^2`.

**Comment by shubhamtibra** [comment]
I subclassed from `MutableDenseMatrix` and overloaded some methods for solver. The `add` and `mul` operation using polynomials only internally and works robust now. In the operation `composition` we need a `subs` method to use polynomials. Eg. substituting x --> x**2 in `DMP( [1, 1], ZZ)` must give `DMP([1, 0, 1], ZZ)` i.e. `x + 1` to `x**2 + 1`.

**Comment by jksuom** [comment]
It seems that there is no ready-made method for that. But it should not be hard to write one following, by example, [Horner's method](https://en.wikipedia.org/wiki/Horner's_method). (A polynomial, or possibly a fraction, is substituted instead of a numerical value.)

File: sympy/holonomic/holonomic.py [hunk 2]
```diff
@@ -325,31 +336,48 @@ def __eq__(self, other):
...
+    num = []
+    denom = []
+    base = parent.base
+    K = base.get_field()
+    R = ZZ.old_poly_ring(base.gens[0])
+    lcm_denom = R.from_sympy(S(1))
+    list_of_coeff = []
+
+    for i, j in enumerate(list_of):
+        if not isinstance(list_of[i], K.dtype):
```
Comment (jksuom) [inline sympy/holonomic/holonomic.py]: It looks like `list_of[i]` is `j`.

Comment (shubhamtibra) [inline sympy/holonomic/holonomic.py]: You are right. The `enumerate` isn't needed. I will change it.

**Comment by jksuom** [comment]
This looks good. You have come quite far in the basic algebraic part.

**Comment by shubhamtibra** [comment]
Thanks @jksuom . What can be the next thing to work on?

**Comment by jksuom** [comment]
At some point Ore algebra of the second kind,  the difference algebra, has to be considered. Maybe this is the time to start thinking of its development.

**Comment by shubhamtibra** [comment]
Yes. You are right. Since its a new feature I think I should open a new PR for that. Is there anything that can be added in this PR?

**Comment by jksuom** [comment]
I cannot think of anything that should be added right now. But probably something will appear later.

**Comment by shubhamtibra** [comment]
I can't think of anything else. This can be reviewed now. Ping @jksuom @certik .

**Comment by shubhamtibra** [comment]
Please review the PR so we can move forward. @certik @jksuom .

File: sympy/holonomic/holonomic.py [hunk 3]
```diff
@@ -434,7 +462,7 @@ def __repr__(self):
             cond_str = ''
             diff_str = ''
             for i in self.y0:
-                cond_str += ', f%s(%d) = %d ' %(diff_str, self.x0, i)
+                cond_str += ', f%s(%s) = %s ' %(diff_str, sstr(self.x0), sstr(i))
```
Comment (kshitij10496) [inline sympy/holonomic/holonomic.py]: A missing whitespace around operator `%`

File: sympy/holonomic/holonomic.py [hunk 4]
```diff
@@ -607,18 +634,21 @@ def __mul__(self, other):
...
+        sol = (NewMatrix(lin_sys).transpose()).gauss_jordan_solve(homo_sys)
         # until a non trivial solution is found
         while sol[0].is_zero:
             # updating the coefficents Dx^i(f).Dx^j(g) for next degree
             for i in range(a - 1, -1, -1):
                 for j in range(b - 1, -1, -1):
                     coeff_mul[i][j + 1] += coeff_mul[i][j]
                     coeff_mul[i + 1][j] += coeff_mul[i][j]
-                    coeff_mul[i][j] = coeff_mul[i][j].diff()
+                    if isinstance(coeff_mul[i][j] , K.dtype):
```
Comment (kshitij10496) [inline sympy/holonomic/holonomic.py]: Extra whitespace

File: sympy/holonomic/holonomic.py [hunk 5]
```diff
@@ -674,6 +702,155 @@ def __mul__(self, other):
...
+        diff = expr.diff()
+        listofpoly = self.annihilator.listofpoly
+
+        for i, j in enumerate(listofpoly):
+            if isinstance(j, self.annihilator.parent.base.dtype):
+                listofpoly[i] = self.annihilator.parent.base.to_sympy(j)
+
+        r = listofpoly[a].subs({self.x:expr})
+        subs = [-listofpoly[i].subs({self.x:expr}) / r for i in range (a)]
+        coeffs = [S(0) for i in range(a)] # Dkfa[i] == coeff of (D^i f)(a) in D^k (f(a))
```
Comment (kshitij10496) [inline sympy/holonomic/holonomic.py]: There should be at least 2 whitespaces before the beginning of an inline comment.

**Comment by kshitij10496** [comment]
@shubhamtibra I have added a few coding style notes.

**Comment by shubhamtibra** [comment]
Thanks @kshitij10496 . I will rectify them at these places and also at other instances where they might not be followed.

**Comment by kshitij10496** [comment]
Cool :+1:

**Comment by shubhamtibra** [comment]
There seem to be a problem with checks. Can anyone restart the build?
I have started working on Recurrence Operators and will open a PR once its done.

**Comment by shubhamtibra** [comment]
Ping @certik @jksuom .

**Comment by jksuom** [comment]
It seems that travis has not even started yet (so there is no way to restart). If it does not start within a reasonable time, you may have to push something. One possibility might be a simplification of the docstrings that are copies from matrices. Maybe a reference to matrices would generally suffice.

File: sympy/holonomic/holonomic.py [hunk 6]
```diff
@@ -699,10 +878,39 @@ def _extend_y0(Holonomic, n):
...
         return y0 + y1[len(y0):]
+
+
+def DMFdiff(frac):
+    if not isinstance(frac, DMF):
+        return frac.diff()
+    K = frac.ring
+    R = K.dom.old_poly_ring(K.gens[0])
+    p = R(frac.num)
+    q = R(frac.den)
```
Comment (jksuom) [inline sympy/holonomic/holonomic.py]: There are methods for these: `K.numer(frac)` and `K.denom(frac)`. `R` is not needed. (There is also a method `R = K.get_ring()`.

File: sympy/holonomic/holonomic.py [hunk 7]
```diff
@@ -674,6 +702,155 @@ def __mul__(self, other):
...
+        tau = sol.atoms(Dummy).pop()
+        sol = sol.subs(tau, 1)
+        sol = _normalize(sol[0:], R, negative=False)
+
+        if self._have_init_cond:
+            return HolonomicFunction(sol, self.x, self.x0, self.y0)
+        return HolonomicFunction(sol, self.x)
+
+
+def From_Hyper(func, x0=0, evalf=False):
```
Comment (certik) [inline sympy/holonomic/holonomic.py]: This should be called `from_hyper`.

**Comment by shubhamtibra** [comment]
Thanks @jksuom @certik . I have made the changes.

**Comment by shubhamtibra** [comment]
The tests have passed. Are we ready to merge it? @certik @jksuom .

**Comment by certik** [comment]
Good job @shubhamtibra, thanks for the work. The functionality is coming along very nicely.

**Pull request merged**